_drafts/arm-cortex-m-exceptions-and-nvic.md

```diff
@@ -0,0 +1,699 @@
+    .long   SPIM1_SPIS1_TWIM1_TWIS1_SPI1_TWI1_IRQHandler
+    .long   NFCT_IRQHandler
+    .long   GPIOTE_IRQHandler
+    .long   SAADC_IRQHandler
+    .long   TIMER0_IRQHandler
+    .long   TIMER1_IRQHandler
+    .long   TIMER2_IRQHandler
+    .long   RTC0_IRQHandler
+    .long   TEMP_IRQHandler
+    .long   RNG_IRQHandler
+    [...]
+```
+
+We can pretty easily compute the **Exception Number** by counting the offset within this table.
+`__StackTop` is **0**, `Reset_Handler` is **1**, `POWER_CLOCK_IRQHandler` is **16**.
+
+Most vendors also
+provide a CMSIS compatible `IRQn_Type` define which also gives you the enumerated list of "External
+Interrupt Numbers" (**Exception Number** - 16). We will want this when we go to configure external
+interrupts that are part of the NVIC. For the NRF52840, this can be found at
+`modules/nrfx/mdk/nrf52840.h` and looks something like this:
+
+```c
+typedef enum {
+[...]
+  POWER_CLOCK_IRQn          =   0,              /*!< 0  POWER_CLOCK                                                            */
+  RADIO_IRQn                =   1,              /*!< 1  RADIO                                                                  */
+  UARTE0_UART0_IRQn         =   2,              /*!< 2  UARTE0_UART0                                                           */
+  SPIM0_SPIS0_TWIM0_TWIS0_SPI0_TWI0_IRQn=   3,  /*!< 3  SPIM0_SPIS0_TWIM0_TWIS0_SPI0_TWI0                                      */
+  SPIM1_SPIS1_TWIM1_TWIS1_SPI1_TWI1_IRQn=   4,  /*!< 4  SPIM1_SPIS1_TWIM1_TWIS1_SPI1_TWI1                                      */
+  NFCT_IRQn                 =   5,              /*!< 5  NFCT                                                                   */
+  GPIOTE_IRQn               =   6,              /*!< 6  GPIOTE                                                                 */
+  SAADC_IRQn                =   7,              /*!< 7  SAADC                                                                  */
+  TIMER0_IRQn               =   8,              /*!< 8  TIMER0                                                                 */
+  TIMER1_IRQn               =   9,              /*!< 9  TIMER1                                                                 */
+  TIMER2_IRQn               =  10,              /*!< 10 TIMER2                                                                 */
+  RTC0_IRQn                 =  11,              /*!< 11 RTC0                                                                   */
+[...]
+  } IRQn_Type;
+```
+
+As discussed above, the actual number of interrupt priority levels is implementation specific. You can
+find the number of levels implemented in the vendors data sheet for the MCU being used or determine it dynamically with
+gdb. Unimplemented bits are Read-as-Zero (RAZ) in the `NVIC_IPR` registers so if we write
+0xff and read it back we can figure out the number of levels. Let's give it a try in GDB:
+
+```
+(gdb) p/x *(uint32_t*)0xE000E400
```

Comment:
Maybe link back to your section about the `NVIC_IPR - 0xE000E400` register?

Reply:
good call